pgjdbc/src/test/java/org/postgresql/test/jdbc2/StatementTest.java

```diff
@@ -411,6 +421,99 @@ public void testWarningsAreCleared() throws SQLException {
+                + "$BODY$ "
+                + "BEGIN "
+                + "RAISE NOTICE 'Test 1'; "
+                + "RAISE NOTICE 'Test 2'; "
+                + "EXECUTE pg_sleep(2); "
+                + "END "
+                + "$BODY$ "
+                + "LANGUAGE plpgsql;");
+    con.createStatement().execute("SET SESSION client_min_messages = 'NOTICE'");
+    final PreparedStatement preparedStatement = con.prepareStatement("SELECT notify_then_sleep()");
+    final AtomicReference<SQLWarning> warning = new AtomicReference<SQLWarning>();
+    ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
+    ScheduledFuture<?> future = executorService.schedule(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          warning.set(preparedStatement.getWarnings());
+        } catch (SQLException e) {
+          fail("Exception thrown: " + e.getMessage());
+        }
+      }
+    }, 1000, TimeUnit.MILLISECONDS);
+    preparedStatement.execute();
+
+    future.get();
+    executorService.shutdown();
+
+    assertNotNull(warning.get());
+    assertEquals(warning.get().getMessage(), "Test 1");
+    assertEquals(warning.get().getNextWarning().getMessage(), "Test 2");
+  }
+
+  /**
+   * Demonstrates a safe approach to concurrently reading the latest
+   * warnings while periodically clearing them.
+   *
+   * One drawback of this approach is that it requires the reader to make it to the end of the
+   * warning chain before clearing it, so long as your warning processing step is not very slow,
+   * this should happen more or less instantaneously even if you receive a lot of warnings.
+   */
+  @Test
+  public void testConcurrentWarningReadAndClear()
+      throws SQLException, InterruptedException, ExecutionException, TimeoutException {
+    final int iterations = 1000;
+    final ExecutorService executor = Executors.newSingleThreadExecutor();
+    final PgStatement statement = (PgStatement) con.createStatement();
+    final Future warningReaderThread = executor.submit(new Callable<Object>() {
+      @Override
+      public Object call() throws SQLException, InterruptedException {
+        SQLWarning lastProcessed = null;
+        int warnings = 0;
+        //For production code replace this with some condition that
+        //ends after the statement finishes execution
+        while (warnings < iterations) {
+          //if next linked warning has value use that, otherwise check latest head
+          final SQLWarning warn = lastProcessed != null && lastProcessed.getNextWarning() != null
+              ? lastProcessed.getNextWarning()
+              : statement.getWarnings();
+          if (warn != null) {
+            warnings++;
+            //System.out.println("Processing " + warn.getMessage());
+            lastProcessed = warn;
+            if (warn == statement.getWarnings()) {
+              //System.out.println("Clearing warnings");
+              statement.clearWarnings();
```

Comment:
I think it would be unlikely someone would replicate it without looking at this unit test.  
Concurrently reading warnings from a statement is probably a fairly niche feature, I think for the majority of use-cases it is unlikely that someone is going to generate enough notices that they would consider using `Statement#clearWarnings()`, but if they do I think the important thing is that the driver doesn’t crash, and we at least provide some way of the user reading the warnings safely, despite how abstruse the implementation might be.  

I considered other approaches like employing locks/synchronization or maybe even exposing a queue data structure, but I felt that those options would introduce negative performance implications for users who don’t need the feature, or in the case of exposing a queue, wouldn’t work with the standard JDBC API.